packages/firestore/test/integration/api/transactions.test.ts

```diff
@@ -20,9 +20,31 @@ import { expect } from 'chai';
 import { Deferred } from '../../util/promise';
 import firebase from '../util/firebase_export';
 import * as integrationHelpers from '../util/helpers';
+import { setAsyncQueue } from '../util/internal_helpers';
+import { AsyncQueue, TimerId } from '../../../src/util/async_queue';
+import { CancelablePromise } from '../../../src/util/promise';
 
-const apiDescribe = integrationHelpers.apiDescribe;
+/**
+ * Version of the AsyncQueue that overrides `enqueueAfterDelay` to fast-forward
+ * the backoffs used in retrying transactions.
```

Comment:
So the whole point of having `TimerId` and a `runDelayedOperationsEarly()` method was so that tests could explicitly fast-forward timers as necessary, _without_ needing to override the AsyncQueue, etc.  So what's the deal here?  Why can't the tests explicitly call `runDelayedOperationsEarly()` as necessary?  Is there some other way we could modify the existing `AsyncQueue` to make transaction retries easily testable?  

I'd like to avoid needing to override / replace the AsyncQueue if at all possible.

Reply:
In case it's useful, an example of a test using runDelayedOperationsEarly(): https://github.com/firebase/firebase-js-sdk/blob/4be253c0443f59dce79c0a4114b5f4a124ce6c06/packages/firestore/test/integration/api_internal/idle_timeout.test.ts#L30

Reply:
The reason I overrode the AsyncQueue was because creating and awaiting the backoff happen inside `runTransaction`. Because of that, once `runTransaction` is called, it will not complete until has completed. How can we separate awaiting the backoff from the rest of the logic?

Reply:
As a somewhat lame solution, if you just did:

setInterval(() => asyncQueue(db).runDelayedOperationsEarly(TimerId.RetryTransaction), 20);

at the start of your test, it seems like that would probably work?  The lame part is we're just calling it incessantly instead of only when a retry backoff is actually enqueued.

I'm not sure what the best solution would be.  Without mocking the network it seems like we can't know when the  retries will be enqueued so we can't know when to call runDelayedOperationsEarly().  So for this kind of test, maybe we need another AsyncQueue test helper method like `queue.overrideDelaysForTimerId(TimerId.RetryTransaction, 0);` to essentially disable the delays.

WDYT?

Reply:
The idea of overriding the delays sounds much more elegant -- wish I could have thought of it sooner! I gave it sho, WDYT?

I'm currently in the process of trying to port this over to android in order to see how the async/await stuff will turn out, but let me know what you think of this approach please